[standalone_compile] support multiple returns

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #151502
* __->__ #151551
* #151501

We were only returning the first one. There's an edge case on what to do
if the original function returns a single Tensor. capture(f) returns a
function that returns a tuple of one Tensor in this case and we were
originally converting this back to one single Tensor. I think it's fine
to return a tuple of one Tensor (that is what the graph passed to
standalone_compile asked for!) but we can revisit.
fine

Test Plan:
- modified one test to used multiple outputs

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

zou3519 committed:
[standalone_compile] support multiple returns

We were only returning the first one. There's an edge case on what to do
if the original function returns a single Tensor. capture(f) returns a
function that returns a tuple of one Tensor in this case and we were
originally converting this back to one single Tensor. I think it's fine
to return a tuple of one Tensor (that is what the graph passed to
standalone_compile asked for!) but we can revisit.
fine

Test Plan:
- modified one test to used multiple outputs

[ghstack-poisoned]

Skylion007 commented on torch/_inductor/standalone_compile.py:
```diff
@@ -58,7 +58,7 @@ def __init__(
         self._artifacts = artifacts
 
     def __call__(self, *args: Any) -> Any:
```
If you use a ParamSpec and a return TypeVar here when you capture the function, you might be able to preserve arg typing and return type here.

zou3519 committed:
Update on "[standalone_compile] support multiple returns"


We were only returning the first one. There's an edge case on what to do
if the original function returns a single Tensor. capture(f) returns a
function that returns a tuple of one Tensor in this case and we were
originally converting this back to one single Tensor. I think it's fine
to return a tuple of one Tensor (that is what the graph passed to
standalone_compile asked for!) but we can revisit.
fine

Test Plan:
- modified one test to used multiple outputs

[ghstack-poisoned]

oulgen commented on test/inductor/test_codecache.py:
```diff
@@ -1486,7 +1486,7 @@ def f(x):
                     path=path, format="unpacked"
                 )
-                compiled_out = loaded(*args)
+                compiled_out = loaded(*args)[0]
```
oh shoot i was completely meaning to fix this before landing, but i must have forgotten, thanks!
